azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/NetworkInterface.java

```diff
@@ -81,35 +81,42 @@
      * @throws CloudException exceptions thrown from the cloud.
      * @throws IOException exceptions thrown from serialization/deserialization.
      */
-    PublicIpAddress publicIpAddress() throws CloudException, IOException;
+    PublicIpAddress primaryPublicIpAddress() throws CloudException, IOException;
 
     /**
      * @return the resource id of the virtual network subnet associated with this network interface.
      */
-    String subnetId();
+    String primarySubnetId();
 
     /**
-     * Gets the virtual network associated with this network interface.
+     * Gets the virtual network associated this network interface's primary Ip configuration.
+     * <p>
+     * note that this method makes a rest API call to fetch the public IP
      *
      * @return the virtual network associated with this network interface.
      * @throws CloudException exceptions thrown from the cloud.
      * @throws IOException exceptions thrown from serialization/deserialization.
      */
-    Network network() throws CloudException, IOException;
+    Network primaryNetwork() throws CloudException, IOException;
 
     /**
-     * Gets the private IP address allocated to this network interface.
+     * Gets the private IP address allocated to this network interface's primary Ip configuration.
      * <p>
      * the private IP will be within the virtual network subnet of this network interface.
      *
      * @return the private IP addresses
      */
-    String privateIp();
+    String primaryPrivateIp();
+
+    /**
+     * @return the private IP allocation method (Dynamic, Static) of this network interface's primary Ip configuration.
+     */
+    String primaryPrivateIpAllocationMethod();
 
     /**
-     * @return the private IP allocation method (Dynamic, Static)
+     * @return the Ip configurations of this network interface
      */
-    String privateIpAllocationMethod();
+    List<? extends NicIpConfiguration> ipConfigurations();
```

Comment:
Subnets and NSGRules child resources does not refer any other resources. NicIpConfig has reference to Subnet and public Ip so afaik we haven't came across a child resource which allow setting reference to other resources.

We could set these reference properties from NicImpl itself, but in that case we are not really taking advantage of  ipconfigImpl class, we endup mixing ipconfig related logic in NicImpl and IpconfigImpl.

we need to discuss and come up with a generic way to handle these type of scenarios.